### ISSUE TITLE
Permanent logging to file

### ISSUE THREAD

**Issue by antoine2711** [OpenRefine/OpenRefine#6686]
Mainly for debugging purposes, I would like to save to files the logging that OR generates.

Currently, you can see it in the terminal, unless you are using the Mac version, and then it’s completly hidden…

This would be very useful, and the only problem that is would had is that it would clutter the disk (not a problem with today’s hard drives size.

### Proposed solution
A new preference of type string, `debugging.logging.folder` or `logging.folder` would have the path to the folder in the back-end (i.e.: "Macintosh HD/Users/antoine/Desktop/OpenRefine Logs", that would store a file per launch of OR. The file would be named by the datetime added to "OpenRefine_" and appended by the file extension ".log", i.e.: `OpenRefine_2024-06-23_010726.log`.

### Alternatives considered
None, but having the user saving the log himself from a copy of the text in the terminal, and no alternative is possible if running the Mac version.

**Comment by antoine2711** [comment]
@tfmorris: I think this should be labelled as a `good first issue`. What do you think?

Regards, Antoine

**Comment by tfmorris** [comment]
This sounds like it would principally be a developer feature. Is that correct? 

If so, I'd look at log4j2.properties. I suspect you could add, for example, a [RollingFileAppender](https://logging.apache.org/log4j/2.x/manual/appenders.html#RollingFileAppender) or some such to the current console appender:
 https://github.com/OpenRefine/OpenRefine/blob/b3329679917cb53235066bee813ecfc9392e5e62/server/src/log4j2.properties#L5-L10

**Comment by GittyHarsha** [comment]
Could you give the exact path to store the logs?
Also since the OR can be saved multiple times in a day
Should we also append  the number of times it was opened to "OpenRefine_YYYY-MM-DD"?

**Comment by antoine2711** [comment]
> Could you give the exact path to store the logs?
The path would be loaded from a preference. 

> Also since the OR can be opened multiple times in a day Should we also append the number of times it was opened to "OpenRefine_YYYY-MM-DD"?

Yes, the time should also be put in the filename, like the example in the description of this issue.

Regards,
   Antoine

**Comment by GittyHarsha** [comment]
We can use a rolling policy based on start-up and size.
1. What can be the maximum size of a log file to trigger a rollover?
2. Should we also keep a policy of deleting old files? If so how many days old files can be deleted?
3. Can we store the old log files in a compressed format?
'

**Comment by antoine2711** [comment]
Hi @GittyHarsha,

> We can use a rolling policy based on start-up and size.
This raise an interesting question, more than just for the log size, but for normal usage too.
If the free space left is minimal (let’s say < 200 Mb), normal operations may hit a point where we are in trouble.
So, as a preventive mesure, we could « seize » at startup 100 Mb, and if we start missing space, then use is to store the current operation and exit. If not, this space could be released at the end of the execution of OR.

@tfmorris: do we have such a mecanism currently to handle « no space left on disk » errors?

> 1. What can be the maximum size of a log file to trigger a rollover?
I would doubt that even if runned on a single exection for more than a week, that the log file would be bigger than 100 Mb.
And since it’s one file per launch, this make a natural limit on each file.

> 2. Should we also keep a policy of deleting old files? If so how many days old files can be deleted?
Hum. If the whole folder of log get to a certain size (let’s say 100 Mb), we could ask the user if he wants to delete them. But I guess we should not ask that at every launch once it as been refused… Maybe store a second preferences saying when to ask the next time (i.e. after asking at 100 Mb), maybe store in `debugging.logging.warnAtSize` the 200 value.

> 3. Can we store the old log files in a compressed format?
I guess it wouldn’t hurt to compress the file once the execution of OR stops.
We could also regroup log in a folder for every months, and when changing month, compress the previous uncompressed months… a bit more complex though.

Regards, Antoine

**Comment by antoine2711** [comment]
> This sounds like it would principally be a developer feature. Is that correct?

@tfmorris: no, actually, it would be to help standard users, because dev can always copy the logs from the terminal, from where they usually run OR.

It would be a good feature for testers, also.

And, like said in the description, for normal Mac user, it would be the only way to get the logs at all.

Regards, Antoine

**Comment by GittyHarsha** [comment]
> A new preference of type string, `debugging.logging.folder` or `logging.folder`

I’m new to the project and was wondering if you could help me understand where the preferences are stored. Any guidance would be greatly appreciated.

**Comment by tfmorris** [comment]
My vision of what error/stack trace logging is different. I think the needs of developers/power users are very different from regular users. For a regular user, I would envision an error modal dialog with a user friendly message ("Something bad happened") and a collapsed pane with the technical details that they could copy/paste -- or a button that they could click to report the error, with associated error log / stack trace, directly. End users shouldn't need to worry about disk quotas, rollover periods, or any of the other things that might be important to a DevOps staff or a developer.

**Comment by GittyHarsha** [comment]
Hi @tfmorris,

I have a question regarding the issue we're addressing:

Is permanent logging always enabled by default? Specifically, should the RollingFileAppender be filtered based on whether the preference string is set or not? If we want to enable or disable permanent logging conditionally based on the availability of a preference string, I believe this isn't feasible using the log4j2.properties file. Instead, it might be achievable with an equivalent XML configuration since filters (e.g., \<Filters>\</Filters>) are supported in XML. For more information, you can refer to the [Log4j Filters documentation](https://logging.apache.org/log4j/2.x/manual/filters.html).

**Comment by antoine2711** [comment]
> Is permanent logging always enabled by default?

No. Only if the preference is set. This will be used by a minority of the users.

> I believe this isn't feasible using the log4j2.properties file. Instead, it might be achievable with an equivalent XML configuration since filters (e.g., <Filters></Filters>) are supported in XML.

Interesting. Thanks for bring this up.

Regards, Antoine

**Comment by antoine2711** [comment]
> For a regular user (…) and a collapsed pane with the technical details that they could copy/paste -- or a button that they could click to report the error, with associated error log / stack trace, directly.

@tfmorris : Yes, I agree that would be much better, but also much more complex to code. My proposition is interesting because it would be easy to implement and would provide an history of problems.

Regards, Antoine

**Comment by magdmartin** [comment]
Is it possible to integrate @tfmorris proposal the JavaFX launcher module #6855 / #3221 that @thadguidry is working on?

**Comment by thadguidry** [comment]
@magdmartin Yes, one of the things that was asked (institutions) was to have easy control of enabling logging to rolling files for troubleshooting.  I plan to offer those log control options in the JavaFX UI, and on the backend, it would enable the rolling file appender.  The config options available are detailed here: https://logging.apache.org/log4j/2.x/manual/appenders/rolling-file.html#recipes

One of the questions in my mind however, was if the JavaFX interface should manipulate and override Environment  variables, or Java System properties, or just do so programmatically https://logging.apache.org/log4j/2.x/manual/customconfig.html.  Programmatically is nice in the sense that it can read and modify `log4j2.xml` files and even allow users to extend (we can ship OpenRefine with sensibly privacy default of "no logging to file", and users can opt-in through checkbox in JavaFX UI for "log to rolling files day/week/month" and "set path of log files").
